float8_experimental/float8_dynamic_linear.py

```diff
@@ -96,8 +107,108 @@ def from_float(cls, mod, emulate: bool = False) -> "Float8DynamicLinear":
                 "bias": False,
             }
             new_mod = cls(**super_kwargs)
-        new_mod.weight = mod.weight
-        new_mod.bias = mod.bias
         new_mod.forward_config = ScaledMMConfig(emulate, True if not emulate else False)
         new_mod.backward_config = ScaledMMConfig(emulate, False)
+        if config.enable_fsdp_fp8_all_gather:
+            new_mod.weight = nn.Parameter(
+                WeightWithDynamicFloat8CastTensor(mod.weight, new_mod.forward_config)
+            )
+        else:
+            new_mod.weight = mod.weight
+        new_mod.bias = mod.bias
         return new_mod
+
+
+# FSDP pads its local tensor on dim-0. The subclass should be preserved such
+# that the padded local tensor (and any transformations like copying to GPU)
+# is of the subclass as well.
+_ops_to_preserve_subclass = {
+    torch.ops.aten.empty_like.default,
+    torch.ops.aten.new_zeros.default,
+    torch.ops.aten.slice.Tensor,
+    torch.ops.aten.copy_.default,
+    torch.ops.aten.view.default,
+    torch.ops.aten.as_strided.default,
+    torch.ops.aten._to_copy.default,
+    torch.ops.aten._pin_memory.default,
+}
+
+
+class WeightWithDynamicFloat8CastTensor(torch.Tensor):
+    @staticmethod
+    def __new__(cls, tensor: torch.Tensor, mm_config: ScaledMMConfig):
+        return torch.Tensor._make_wrapper_subclass(
+            cls,
+            tensor.size(),
+            strides=tensor.stride(),
+            storage_offset=tensor.storage_offset(),
+            memory_format=suggest_memory_format(tensor),
+            dtype=tensor.dtype,
+            layout=tensor.layout,
+            device=tensor.device,
+            pin_memory=tensor.is_pinned(),
+            requires_grad=tensor.requires_grad,
+        )
+
+    def __init__(self, tensor: torch.Tensor, mm_config: ScaledMMConfig):
+        self._tensor = tensor
+        self._mm_config = mm_config
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args, kwargs=None):
+        if func == torch.ops.aten.detach.default:
+            return WeightWithDynamicFloat8CastTensor(
+                args[0]._tensor, args[0]._mm_config
+            )
+        mm_config: Optional[ScaledMMConfig] = None
+
+        def unwrap(t):
+            nonlocal mm_config
```

Comment:
trying to wrap my brain around this...if mm_config is not coming from `t`, then where is it coming from?  do we ever need to merge it other than for scaled_mm?

Reply:
I think this was more for future proofing. As is, since we only have a select few ops preserve the subclass and none of those ops take in >1 tensor input, then we should never have to merge. However, in the future, we may want to have more ops preserve subclass.

I can remove the merge logic and error with `NotImplementedError` if you prefer. Feel free to let me know if this does not make sense.

---

> if mm_config is not coming from t, then where is it coming from?

I think `mm_config` is always coming from the `t`s in the inputs.